monitor/services/src/main/java/org/datacleaner/monitor/configuration/ComponentCacheMapImpl.java

```diff
@@ -0,0 +1,241 @@
+        }
+        componentCacheConfigWrapper.updateTimeStamp();
+        return componentCacheConfigWrapper;
+    }
+
+    /**
+     * Remove configuration from memory and store. And this component is destroyed.
+     *
+     * @param id
+     * @param tenantContext
+     */
+    public boolean remove(String id, TenantContext tenantContext) {
+        boolean inCache = removeConfigurationOnlyFromCache(id);
+        boolean inStore = removeConfigurationOnlyFromStore(id, tenantContext);
+        logger.info("Component {} was removed from cache and closed.", id);
+        return inCache || inStore;
+    }
+
+
+    private boolean removeConfigurationOnlyFromCache(String id) {
+        ComponentCacheConfigWrapper config = data.get(id);
+        if (config != null) {
+            data.remove(id);
+            config.getHandler().closeComponent();
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private boolean removeConfigurationOnlyFromStore(String id, TenantContext tenantContext) {
+        ComponentStore store = tenantContext.getComponentStore();
+        return store.remove(id);
+    }
+
+    /**
+     * Close all component in memory. All components configuration are still in repository.
+     *
+     * @throws InterruptedException
+     */
+    public void close() throws InterruptedException {
+        logger.info("Closing Components cache.");
+
+        synchronized (checker) {
+            checker.stop();
+            checker.notifyAll();
+        }
+        long maxTime = System.currentTimeMillis() + CLOSE_TIMEOUT;
+        while (checkerThread.isAlive()) {
+            Thread.sleep(500);
+            if (maxTime < System.currentTimeMillis()) {
+                logger.error("Problem with closing checking thread.");
+                break;
+            }
+        }
+
+        for (ComponentCacheConfigWrapper componentCacheConfigWrapper : data.values()) {
+            componentCacheConfigWrapper.getHandler().closeComponent();
+            logger.info("Component with id: {} was closed.", componentCacheConfigWrapper.getComponentStoreHolder().getInstanceId());
+            // Configuration is still in store.
+        }
+
+        data.clear();
+        logger.info("Components cache was closed.");
+    }
+
+
+    /**
+     * Thread for checking timeout for each components and also do update configuration is store.
+     */
+    private class TimeoutChecker implements Runnable {
+        boolean running = true;
+        boolean firstRun = true;
+
+        @Override
+        public void run() {
+            while (running) {
+                if (firstRun) {
+                    firstRun = false;
+                } else {
+                    Iterator<RepositoryFolder> repositoryFolderIterator = _tenantContextFactory.getRepositoryFolderIterator();
+                    Set<String> allIdInCache = new HashSet<>(data.keySet());
+
+                    while (repositoryFolderIterator.hasNext()) {
```

Comment:
The reason for cycling through all the tenants is a potential problem in load balancing. When nodeA goes down after some time of work it has some repositories stored in the shared file system but the content of the cache (in memory) is gone. NodeB then has to take care of the "lost" repositories. 

If this is a really big problem we could think of another solution: let's not use tenants' directories but a special new shared directory for storing component configurations. They have an unique ID. I am unaware of the complexity of this but I can certainly try it. 

Or do you have a better idea?